### DISCUSSION TITLE
How to generate generic templates

### DISCUSSION THREAD

**Discussion by amoscatelli** [eclipse/jkube#2902] [Q&A]
Hi everybody,
is it possible to generate a generic template/helm chart by using k8s:resources goal ?

For example by using "helm create" something similar is created for ingress : 

```
{{- if .Values.ingress.enabled -}}
{{- $fullName := include "chart.fullname" . -}}
{{- $svcPort := .Values.service.port -}}
{{- if and .Values.ingress.className (not (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion)) }}
  {{- if not (hasKey .Values.ingress.annotations "kubernetes.io/ingress.class") }}
  {{- $_ := set .Values.ingress.annotations "kubernetes.io/ingress.class" .Values.ingress.className}}
  {{- end }}
{{- end }}
{{- if semverCompare ">=1.19-0" .Capabilities.KubeVersion.GitVersion -}}
apiVersion: networking.k8s.io/v1
{{- else if semverCompare ">=1.14-0" .Capabilities.KubeVersion.GitVersion -}}
apiVersion: networking.k8s.io/v1beta1
{{- else -}}
apiVersion: extensions/v1beta1
{{- end }}
kind: Ingress
metadata:
  name: {{ $fullName }}
  labels:
    {{- include "chart.labels" . | nindent 4 }}
  {{- with .Values.ingress.annotations }}
  annotations:
    {{- toYaml . | nindent 4 }}
  {{- end }}
spec:
  {{- if and .Values.ingress.className (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion) }}
  ingressClassName: {{ .Values.ingress.className }}
  {{- end }}
  {{- if .Values.ingress.tls }}
  tls:
    {{- range .Values.ingress.tls }}
    - hosts:
        {{- range .hosts }}
        - {{ . | quote }}
        {{- end }}
      secretName: {{ .secretName }}
    {{- end }}
  {{- end }}
  rules:
    {{- range .Values.ingress.hosts }}
    - host: {{ .host | quote }}
      http:
        paths:
          {{- range .paths }}
          - path: {{ .path }}
            {{- if and .pathType (semverCompare ">=1.18-0" $.Capabilities.KubeVersion.GitVersion) }}
            pathType: {{ .pathType }}
            {{- end }}
            backend:
              {{- if semverCompare ">=1.19-0" $.Capabilities.KubeVersion.GitVersion }}
              service:
                name: {{ $fullName }}
                port:
                  number: {{ $svcPort }}
              {{- else }}
              serviceName: {{ $fullName }}
              servicePort: {{ $svcPort }}
              {{- end }}
          {{- end }}
    {{- end }}
{{- end }}
```

So this is almost completely driven by values.yaml.
By using the k8s:resources goal can I achieve something similar ?

I don't want to specify such details during the maven compilation, I want to specify things like hostnames or tsl secret name later.

Thank you in advance

**Comment by manusa** [comment]
Not with such a complicated template (if statements, etc.).

However you can achieve a certain level of complexity by combining values and go templates, see the [spring-boot-helm](https://github.com/eclipse/jkube/blob/30379f750e4ea87147c618e612ce1a79d4cd4f96/quickstarts/maven/spring-boot-helm) quickstart for details.

There's also this presentation from EclipseCon 2023 which describes how JKube works (at the moment) regarding Helm: https://www.youtube.com/watch?v=DvMickoLlXM

Notice the difference between the inner and outer loop cycles.

k8s:resource for inner loop needs to be able to generate valid YAML files that can be applied directly to the cluster

k8s:resource + k8s:helm for the outer loop needs to be able to generate the Helm templates with the appropriate template content.

This brings in a set of challenges that we're still trying to overcome.
See for example:
- https://github.com/eclipse/jkube/pull/2689
- https://github.com/eclipse/jkube/issues?q=is%3Aopen+is%3Aissue+label%3A%22component%2Fhelm+%E2%8E%88%22

**Comment by amoscatelli** [comment]
Thank you for your answer.
Is it possible to achieve the same without making any changes to pom.xml ?

What I don't like is to specificy every substitution inside pom.xml : 
https://github.com/eclipse/jkube/blob/master/quickstarts/maven/spring-boot-helm/pom.xml

I tried to use helm variables directly but I got a : 
"spec.replicas: string found, integer expected, violation type=type"

**Reply by manusa** [reply]
You could use a template file for that: https://github.com/eclipse-jkube/jkube/blob/729dc8a0e201499b7fde995d7e61a7de8da110ae/quickstarts/maven/spring-boot-helm/src/main/jkube/template.yaml

However, the recommended approach is to use pom.xml instead.

Check the [quickstart](https://github.com/eclipse-jkube/jkube/tree/729dc8a0e201499b7fde995d7e61a7de8da110ae/quickstarts/maven/spring-boot-helm) to see how some of this variables defined as template parameters are used in this example (for example the deployment's resource limits).
